wrc/codegen/merger.py

```diff
@@ -0,0 +1,177 @@
+from wrc.sema.ast import Article, Section, TableOfContent, Guideline, split_rule_number
+from copy import deepcopy
+
+NOTES_INDEX = 0
+TOC_INDEX = 1
+FIRST_ARTICLE_INDEX = 2
+
+
+class BadFormatError(Exception):
+    def __init__(self, message: str):
+        self.message = message + '. Please report this issue to the developer.'
+        super().__init__(message)
+
+
+def compare_article_numbers(num1: str, num2: str) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    This is necessary because:
+    * We need to compare article numbers like '1' and 'A'.
+    * '2' > '10' == True, which is a problem.
+    """
+    is_digit_1 = num1.isdigit()
+    is_digit_2 = num2.isdigit()
+    if is_digit_1 and is_digit_2:
+        res = int(num1) < int(num2)
+    elif is_digit_1 and not is_digit_2:
+        # Only num2 is a letter.
+        res = True
+    elif not is_digit_1 and not is_digit_2:
+        res = num1 < num2
+    else:
+        res = False
+
+    return res
+
+
+def compare_reg_numbers(num1: list, num2: list) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    We need this for the same reasons we need compare_article_numbers().
+    """
+    for n1, n2 in zip(num1, num2):
+        if n1 < n2:
+            return True
+        elif n1 > n2:
+            return False
+
+    return len(num1) < len(num2)
+
+
+def get_reg_number(reg):
+    if isinstance(reg, Guideline):
+        reg = reg.number.split('+')[0]
+        num = split_rule_number(reg)
+    else:
+        num = split_rule_number(reg.number)
+    if str(num[-1]) == '0':
+        # Because split_rule_number() sometimes returns a 0 at the end, and we don't want that here.
+        del num[-1]
+    return num
+
+
+def merge_ast(astreg, astguide, language_options):
+    """
+    Combine the two AST into one.
+    """
+    # We only want the Articles, so we will start the for loop below at FIRST_ARTICLE_INDEX,
+    # and check if the indices for Notes, TOC and First Article are correct.
+    if not isinstance(astreg.sections[NOTES_INDEX], Section) or \
+            not isinstance(astguide.sections[NOTES_INDEX], Section):
+        raise BadFormatError('NOTES_INDEX is incorrect')
+    if not isinstance(astreg.sections[TOC_INDEX], TableOfContent):
+        raise BadFormatError('TOC_INDEX is incorrect')
+    if not (hasattr(astreg.sections[FIRST_ARTICLE_INDEX], 'number') and astreg.sections[
+                FIRST_ARTICLE_INDEX].number == '1'):
+        raise BadFormatError('FIRST_ARTICLE_INDEX is incorrect.')
```

Comment:
Uber nitpick: This ends in a fullstop `.` and the __init__ of `BadFormatError` will append another `.`.

Reply:
Good catch! Fixed.